test/e2e/operator_test.go

```diff
@@ -442,3 +444,114 @@ func TestDNSForwarding(t *testing.T) {
+		},
+	}
+	if err := cl.Update(context.TODO(), defaultDNS); err != nil {
+		t.Fatalf("failed to update dns %s: %v", defaultDNS.Name, err)
+	}
+	defer func() {
+		defaultDNS = &operatorv1.DNS{}
+		if err := cl.Get(context.TODO(), name, defaultDNS); err != nil {
+			t.Fatalf("failed to get default dns: %v", err)
+		}
+		if defaultDNS.Spec.NodePlacement.NodeSelector != nil {
+			defaultDNS.Spec.NodePlacement.NodeSelector = nil
+			if err := cl.Update(context.TODO(), defaultDNS); err != nil {
+				t.Fatalf("failed to update dns %s: %v", defaultDNS.Name, err)
+			}
+		}
+	}()
+
+	// Verify that all remaining DNS pods are running on master nodes.
+	dnsDaemonSet := &appsv1.DaemonSet{}
+	dnsDaemonSetName := operatorcontroller.DNSDaemonSetName(defaultDNS)
+	if err := cl.Get(context.TODO(), dnsDaemonSetName, dnsDaemonSet); err != nil {
+		fmt.Errorf("failed to get daemonset %s: %v", dnsDaemonSetName, err)
+	}
+	selector, err := metav1.LabelSelectorAsSelector(dnsDaemonSet.Spec.Selector)
+	if err != nil {
+		t.Fatalf("daemonset %s has invalid spec.selector: %v", dnsDaemonSetName, err)
+	}
+	listOpts := []client.ListOption{
+		client.MatchingLabelsSelector{Selector: selector},
+		client.InNamespace(dnsDaemonSet.Namespace),
+	}
+	nodeList := &corev1.NodeList{}
+	if err := cl.List(context.TODO(), nodeList); err != nil {
+		t.Fatalf("failed to list nodes: %v", err)
+	}
+	masterNodes := sets.NewString()
+	for _, node := range nodeList.Items {
+		if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
+			masterNodes.Insert(node.Name)
+		}
+	}
+	podList := &corev1.PodList{}
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		if err := cl.List(context.TODO(), podList, listOpts...); err != nil {
+			t.Logf("failed to list pods for dns daemonset %s: %v", dnsDaemonSetName, err)
+			return false, nil
+		}
+		for _, pod := range podList.Items {
+			if !masterNodes.Has(pod.Spec.NodeName) {
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		nodes := sets.NewString()
+		for _, node := range nodeList.Items {
+			nodes.Insert(node.Name)
+		}
+		t.Errorf("failed to observe updated set of nodes with dns pods; expected %s; got %s", strings.Join(masterNodes.List(), ","), strings.Join(nodes.List(), ","))
+	}
+
+	// Configure DNS pods to run only on nodes with label foo=bar (which we
+	// assume no nodes have).
+	if err := cl.Get(context.TODO(), name, defaultDNS); err != nil {
+		t.Fatalf("failed to get default dns: %v", err)
+	}
+	defaultDNS.Spec.NodePlacement.NodeSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"foo": "bar",
+		},
+	}
+	if err := cl.Update(context.TODO(), defaultDNS); err != nil {
+		t.Fatalf("failed to update dns %s: %v", defaultDNS.Name, err)
+	}
+
+	// Verify that DNS pods continue to run (meaning the label selector that
+	// matched no nodes is being ignored).
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
```

Comment:
Might make sense to run this polling loop for more than one minute, since if the DNS daemonset were to scale down to zero replicas, that would presumably take some time.

Reply:
Good point.  I don't want  to slow down CI's happy path too much, but it would make sense to wait a few minutes.  I've increase the polling loop's timeout to 3 minutes.  I've also increased the timeout of the other polling loop (that verifies that pods are only running on master nodes when a selector for master nodes has been set) to 3 minutes to reduce the risk that that loop could cause flakiness.